### PR TITLE
batched processing

### PR DESCRIPTION
**Pull request by dkrako**
This is a work in progress for implementing the following issue:
[use batched processing instead of processing by instance #80 ](https://github.com/understandable-machine-intelligence-lab/Quantus/issues/80)

### CODE REVIEW

**Comment by sebastian-lapuschkin** [comment]
Batch processing will, given sufficient data samples to be processed, most likely improve throughput and performance (always depending on the system).
Batching the data along the sample axis of the given data has its limitations in terms of flexibility.
Consider the case where a user wants to perform perturbation analysis where single pixels are replaced and the perturbed sample is to be evaluated by the model, however only for a single data point. In this case, no performance benefits will apply as in-parallel perturbation of all available samples will result in a sequential evaluation.

I therefore propose the following approach, which does not create data batches according to the given samples to be evaluated, but in terms of  "work packages" resulting from all to be evaluated combinations of data samples times perturbation steps. In such a manner, the perturbations could be computed and prepared before the perturbed samples will be passed through the network, in almost arbitrary order (similar as to a data generator for NN training would operate), and the delta f(x) for each evaluated instance could be collected easily. This way, even when single samples are to be perturbed numerous times, those perturbed states could be packaged into neural network input batches and evaluated more efficiently.
That is, I think it can safely be assumed that the input level perturbations, even though the individual perturbation steps usually are not independent from each other (in most cases we want to perform successive data perturbation building upon the previously perturbed sample state), will be way less computationally expensive than a pass through the network.

consider the following pseudo code:

```
net: a neural network function
X:  data samples
A : attributions corresponding to X
k : nn_input_batch_size

# use a worker pool to smartly prepare the perturbed samples. 
# creates n workers in an e.g. cpu-bound worker pool (the user should specify n here)
workers = create_workers(n)
with workers:
     for perturbed_samples in generate_perturbations(workers, X, A, k):
        pred = net.forward()
        [...] do stuff with pred
```

the basic idea is to use `generate_perturbations` to prepare some k-sized batches using n workers from X according t A. Since a lot of combinations of datapoints times perturbation states will be generated overall,  generate_perturbations could internally work with a buffer of a low number of pre-computed batches to assure a good and fluid data transfer between the worker output and the neural network. Further, `generate_perturbations` must and should not be bound to parallelize over the data axis of X, which would limit applicability again in the single-sample case, but just return the next k perturbed states from datapoints in X, ie also parallelize over the perturbation state axis.

**Comment by sebastian-lapuschkin** [comment]
bump, pseudocode update

**Comment by leanderweber** [comment]
Just noting this, so we remember: PR #88 contains some indexing functions which are tailored to the current non-batching implementation of metrics in the main branch. These functions would also need to be (slightly) altered for batched processing. I have left some TODOs w.r.t. that in the code also

File: quantus/helpers/pytorch_model.py [hunk 1]
```diff
@@ -23,26 +27,43 @@ def predict(self, x, softmax_act=False, **kwargs):
...
+        predict_kwargs = self.predict_kwargs
+        #device = predict_kwargs.pop("device", None)
+        # TODO: remove this as it depends on pytorch lightning
+        device = self.model.device
+
         grad_context = torch.no_grad() if not grad else suppress()
 
         with grad_context:
-            pred = self.model(torch.Tensor(x).to(device))
+            pred = self.model(torch.Tensor(x).to(device))#, **predict_kwargs)
```
Comment (leanderweber) [inline quantus/helpers/pytorch_model.py:41]: commented out?

File: quantus/helpers/utils.py [hunk 2]
```diff
@@ -343,21 +376,41 @@ def get_nr_patches(
...
 
-def _pad_array(arr: np.array, pad_width: int, mode: str, omit_first_axis=True):
+def pad_array(arr: np.ndarray, pad_width: int, mode: str,
+              omit_first_axis: bool = True, batched: bool = True,
+) -> np.ndarray:
     """To allow for any patch_size we add padding to the array."""
     pad_width_list = [(pad_width, pad_width)] * arr.ndim
-    if omit_first_axis:
+    
+    if batched:
```
Comment (leanderweber) [inline quantus/helpers/utils.py:385]: the batched boolean (everywhere that it is employed) feels a bit weird.

I think a lot less cluttered solution would be to simply always introduce a batch axis during preprocessing if it is not yet present, and then work based on that in all functions.

File: quantus/helpers/utils.py [hunk 3]
```diff
@@ -59,35 +62,39 @@ def get_baseline_value(
...
+                         "neighbourhood_mean", "neighbourhood_random_min_max"]
+        if choice not in valid_choices:
+            raise ValueError(f"Ensure that 'choice'(str) is in {valid_choices}")
+        elif choice == "mean":
+            return float(arr.mean())
+        elif choice == "random":
+            return float(random.random())
+        elif choice == "uniform":
+            return float(np.random.uniform(arr.min(), arr.max()))
+        elif choice == "black" or choice == "min":
```
Comment (leanderweber) [inline quantus/helpers/utils.py:76]: should "black" not result in a 0 ?

File: quantus/helpers/utils.py [hunk 4]
```diff
@@ -59,35 +62,39 @@ def get_baseline_value(
...
+            return float(arr.min())
+        elif choice == "white" or choice == "max":
```
Comment (leanderweber) [inline quantus/helpers/utils.py:78]: similarly, "white" should be a 1 or 255, right?

**Review by leanderweber — CHANGES_REQUESTED** [review]
Noticing this mid-review, from perturb_func.py and utils.py: Changes seem based on a months-old version of the main branch. Please merge the current main branch into here first, since a lot of functionality changed since then.

**Comment by annahedstroem** [comment]
An update version of this work can be found in: https://github.com/understandable-machine-intelligence-lab/Quantus/pull/168.
